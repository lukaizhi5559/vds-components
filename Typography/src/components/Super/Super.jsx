import styled from 'styled-components';
import React from 'react';
import PropTypes from 'prop-types';
import { Fonts } from '../../fonts';

const propTypes = {
  /**
   * If true, custom superscript styles will be applied
   */
  superscriptSmall: PropTypes.bool,
  /**
   * @ignore
   */
  feature: PropTypes.bool,
  /**
   * Assigns an testid to the component
   */
  'data-testid': PropTypes.string,
  /**
   * String or number that is used to render as the superscript
   */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

const StyledSuper = styled.sup`
  ${({ feature, size }) =>
    feature &&
    `
    position: relative;
    font-size: ${size / 2}px;
    line-height: 0px;
    font-family: ${Fonts.VerizonNHGeDS};
  `};
`;

const Super = props => {
  const {
    fontSize,
    children,
    superscriptSmall,
    feature,
    'data-testid': testId,
  } = props;
  const stringifiedChildren = children.toString();
  const size = fontSize && parseInt(fontSize, 10);
  let Wrapper, superProps;
  /**
   * This is to avoid adding size and feature when the element is a span
   * "invalid dom attribute of..."/ transient prop error
   */
  if (superscriptSmall) {
    Wrapper = StyledSuper;
    superProps = {
      size,
      feature,
    };
  } else {
    Wrapper = 'span';
    superProps = {};
  }

  function _convertToUnicode(char, i) {
    char = char.toString();
    switch (char) {
      case '0':
        return '\u2070';
      case '1':
        return '\u00B9';
      case '2':
        return '\u00B2';
      case '3':
        return '\u00B3';
      case '4':
        return '\u2074';
      case '5':
        return '\u2075';
      case '6':
        return '\u2076';
      case '7':
        return '\u2077';
      case '8':
        return '\u2078';
      case '9':
        return '\u2079';
      default:
        return superscriptSmall ? char : <sup key={i}>{char}</sup>;
    }
  }

  function _parseChildren(children) {
    return [...children].map((char, i) => {
      return _convertToUnicode(char, i);
    });
  }

  if (typeof children !== 'string' && typeof children !== 'number') {
    console.warn(
      'Super element only allows children of type string or number.'
    );
    return null;
  }

  return (
    <Wrapper {...superProps} data-testid={testId}>
      {_parseChildren(stringifiedChildren)}
    </Wrapper>
  );
};

Super.propTypes = propTypes;
Super.displayName = 'Super';
export default Super;
