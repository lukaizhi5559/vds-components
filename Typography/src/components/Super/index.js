export { default as Super } from './Super';
