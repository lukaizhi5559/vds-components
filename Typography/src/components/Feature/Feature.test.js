import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { Feature } from '.';
import { Super } from '../Super';
import { TypographyTokens } from '@vds-tokens/typography';
import { ColorTokens } from '@vds-tokens/color';

function _calcSize(remVal) {
  const baseFont = 16;
  let num = parseFloat(remVal);
  return num * baseFont + 'px';
}

function componentToHex(c) {
  var hex = c.toString(16);
  return hex.length == 1 ? '0' + hex : hex;
}

function rgbToHex(rgb) {
  rgb = rgb
    .substr(4)
    .split(')')[0]
    .split(', ');
  return (
    '#' +
    componentToHex(rgb[0]) +
    componentToHex(rgb[1]) +
    componentToHex(rgb[2])
  );
}

test('should render Feature default props correctly', () => {
  const content = 'Feature default';
  const { getByText, container } = render(<Feature>{content}</Feature>);
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  fireEvent.click(container.firstChild);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;

  // should match with default size medium
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.feature[96].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.feature[88].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.light.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});

// Test passing a different color
test('should render Feature white correctly', () => {
  const content = 'Feature white';
  const { getByText, container } = render(
    <Feature color="#FFFFFF">{content}</Feature>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(styles.color).toBe('rgb(255, 255, 255)');
  }
});

//Test passing a different primitive
test('should render Feature primitive as span', () => {
  const content = 'Feature span';
  const { getByText, container } = render(
    <Feature primitive={'span'}>{content}</Feature>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  const elem = container.querySelector('span');
  expect(elem).toBeInTheDocument();
});

// Test passing a different size
test('should render Feature large correctly', () => {
  const content = 'Feature large';
  const { getByText, container } = render(
    <Feature size={'large'}>{content}</Feature>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.feature[128].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.feature[120].value
    );
  }
});

test('should render Feature XLarge as XLarge', () => {
  const content = 'Feature large';
  const { getByText, container } = render(
    <Feature size={'XLarge'}>{content}</Feature>
  );

  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.feature[144].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.feature[136].value
    );
  }
});

// Test passing a superscriptSmall
test('should render Feature with superscriptSmall prop and Super', () => {
  const content = 'Feature large';
  const { getByText, container } = render(
    <Feature size={'large'} superscriptSmall>
      {content}
      <Super>1</Super>
    </Feature>
  );

  expect(container.firstChild).toMatchSnapshot();
  const elem = container.querySelector('sup');
  expect(elem).toBeInTheDocument();
  //Find styling props
  const stylesFeature = window.getComputedStyle(container.firstChild)._values;
  const stylesSuper = window.getComputedStyle(elem)._values;
  if (stylesFeature) {
    expect(_calcSize(stylesFeature['font-size'])).toBe(
      TypographyTokens.fontsize.feature[128].value
    );
    expect(_calcSize(stylesFeature['line-height'])).toBe(
      TypographyTokens.lineheight.feature[120].value
    );
  }
  if (stylesSuper) {
    expect(stylesSuper['font-size']).toBe('64px');
    expect(stylesSuper['line-height']).toBe('0px');
  }
});

// Test passing a bold false
test('should render Feature bold false correctly', () => {
  const content = 'Feature bold false';
  const { getByText, container } = render(
    <Feature size={'large'} bold={false}>
      {content}
    </Feature>
  );

  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.light.value
    );
  }
});

// Test passing a different fontWeight
test('should render Feature fontWeight 750 correctly', () => {
  const content = 'Feature fontWeight 750';
  const { getByText, container } = render(
    <Feature size={'large'} bold={true}>
      {content}
    </Feature>
  );

  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.bold.value
    );
  }
});

// Test changing the viewport
test('should render Feature with default styles on mobile correctly', () => {
  const content = 'Feature on mobile';
  const { getByText, container } = render(
    <Feature viewport={'mobile'} bold={true}>
      {content}
    </Feature>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.feature[64].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.feature[64].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.bold.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
  }
});

// Test passing a different fontSize and lineheight
test('should render Feature fontSize and lineheight props correctly', () => {
  const content = 'Feature fontSize 16 and lineheight 16';
  const { getByText, container } = render(
    <Feature fontSize={16} lineheight={16}>
      {content}
    </Feature>
  );

  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.feature[96].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.feature[88].value
    );
  }
});

// Test Feature without children
test('should render Feature without children', () => {
  const { container } = render(<Feature></Feature>);
  expect(container.firstChild).toMatchSnapshot();
});

// Test Feature without config
test('should not render Feature without config', () => {
  const content = 'Feature will not be rendered without config';
  const { container } = render(<Feature>{content}</Feature>);
  expect(container.firstChild).toBeNull;
});
