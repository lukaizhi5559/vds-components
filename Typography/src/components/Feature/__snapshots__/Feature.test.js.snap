// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Feature default props correctly 1`] = `
.c1 {
  font-size: 6rem;
  font-weight: 300;
  line-height: 5.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<h4
  aria-hidden="false"
  class="c0 c1"
>
  Feature default
</h4>
`;

exports[`should render Feature large correctly 1`] = `
.c1 {
  font-size: 8rem;
  font-weight: 300;
  line-height: 7.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<h3
  aria-hidden="false"
  class="c0 c1"
>
  Feature large
</h3>
`;

exports[`should render Feature primitive as span 1`] = `
.c1 {
  font-size: 6rem;
  font-weight: 300;
  line-height: 5.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<span
  aria-hidden="false"
  class="c0 c1"
>
  Feature span
</span>
`;

exports[`should render Feature white correctly 1`] = `
.c1 {
  font-size: 6rem;
  font-weight: 300;
  line-height: 5.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #FFFFFF;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<h4
  aria-hidden="false"
  class="c0 c1"
>
  Feature white
</h4>
`;

exports[`should render Feature with default styles on mobile correctly 1`] = `
.c1 {
  font-size: 4rem;
  font-weight: 700;
  line-height: 4rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<h4
  aria-hidden="false"
  class="c0 c1"
>
  Feature on mobile
</h4>
`;

exports[`should render Feature with superscriptSmall prop and Super 1`] = `
.c1 {
  font-size: 8rem;
  font-weight: 300;
  line-height: 7.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

.c2 {
  position: relative;
  font-size: 64px;
  line-height: 0px;
  font-family: Verizon-NHG-eDS;
}

<h3
  aria-hidden="false"
  class="c0 c1"
>
  Feature large
  <sup
    class="c2"
    size="128"
  >
    ยน
  </sup>
</h3>
`;

exports[`should render Feature without children 1`] = `
.c1 {
  font-size: 6rem;
  font-weight: 300;
  line-height: 5.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:hover {
  outline: none;
}

<h4
  aria-hidden="false"
  class="c0 c1"
/>
`;
